### ISSUE TITLE
Different behavior for nested generic functions between JSDoc and TS declarations

### ISSUE THREAD

**Issue by IllusionMH** [microsoft/TypeScript#36494]
**TypeScript Version:**  3.7.5 and 3.8.0-dev.20200125

**Search Terms:**
jsdoc generics, jsdoc nested functions, jsdoc nested generic functions, jsdoc curried functions

**Code**
JS Code that doesn't report an error
```js
/** @type {<T>(a: T) => (b: T) => T} */
const curriedFunction = a => b => b;

const text1 = "hello";
const text2 = "Mattia";

const text3 = curriedFunction(text1)(text2);
const text4 = curriedFunction(text1)(123); // no error in JS
```
TS equivalent reports an error
```ts
const curriedFunction: <T>(a: T) => (b: T) => T = a => b => b;

const text1 = "hello";
const text2 = "Mattia";

const text3 = curriedFunction(text1)(text2);
const text4 = curriedFunction(text1)(123); // error in TS
```

However if type is declared in `helpers.d.ts` and used in JS file - it works as expected.
```ts
export type Curried = <T>(a: T) => (b: T) => T;
```
```js
/** @type {import('./helpers').Curried} */
const curriedFunction = a => b => b;

const text1 = "hello";
const text2 = "Mattia";

const text3 = curriedFunction(text1)(text2);

const text4 = curriedFunction(text1)(123); // reports error in JS file now
```

**Expected behavior:**
Hover type info looks similar for JSDoc and TS annotations.
Return type correctly matches type of first argument.
**Error reported when number passed to second call** - same behavior between TS, JS, and JS + .d.ts.

**Actual behavior:**
Hover type info looks similar for JSDoc and TS annotations.
Return type correctly matches type of first argument.
**No error when number passed to second call** - behavior differs form TS and JS + .d.ts.

**Playground Link:**
[JS example](https://www.typescriptlang.org/play/index.html?strictNullChecks=true&ts=3.8.0-dev.20200125&useJavaScript=true#code/PQKhAIAEBcE8AcCm4DeAeAKgPgBQEMAucDASnAF4twcAjI0iqjAX3BGAFgAoAYwHsAdgGdo4HgFcATpICWiACYAxcQJ7QZgiuDyNwNXTQDc3bv2GjoiAB7QAjFoBEAC0QAbV3wfHegkeEs2AEyOALJ40Op4XiY+5v7W0ADMWhLSckoqahoCOAF2JLkJgSTeZn55ACwpUrIKyqrqgoU2tgW2gYkl4MDA4AJ84IjSfJLgMgLgAFIAykA)
[TS example](https://www.typescriptlang.org/play/index.html?ts=3.8.0-dev.20200125#code/MYewdgzgLgBMCuAnRBLApgEwGLzMKK4AXDADwAqAfABQCGJ5AlDALyUzUBGDzbM5rGLVbtOImJwDcAWABQc0JFhQ0ADygBGQQCIAFmgA2BkNpmzF0GCvUAmHQFlaUArVNyF4S9agBmQQmR0bFx8QjBqbw1GCLUoG0YzC2VYgBZ-JFRMHDwCcBj1KOoNGx8EmAB6cpg0ZBBEGBQwfgBlIA)

**Related Issues:** 
Discovered as part of #36201
/cc @sandersn (sorry for delay with this issue)

**Comment by IllusionMH** [comment]
Fun fact - Playground v3 beta shows error in JS variant and VSCode Insiders now also shows an error.

Will check if there any other cases when this one is reproducible somewhere except Playground v2

**Comment by ajafff** [comment]
> Will check if there any other cases when this one is reproducible somewhere except Playground v2

I think this is an issue with the current playground. It doesn't seem to show errors for JS files.

**Comment by IllusionMH** [comment]
@ajafff that's my assumption too.
I checked other cases in VS Code and tsc and it doesn't look like that TS won't report error in this case if `checkJS` or `// @ts-check` are used.

Opened https://github.com/microsoft/TypeScript-Website/issues/203 to see if it supposed to work in V2 or we should wait for V3?
